sui_programmability/examples/regulated_coin/sources/RegulatedCoin.move

```diff
@@ -0,0 +1,575 @@
+        id: VersionedID,
+        balance: Balance<T>,
+        owner: address
+    }
+
+    /// Get the `RegulatedCoin.balance.value` field;
+    public fun value<T>(c: &RegulatedCoin<T>): u64 {
+        Balance::value(&c.balance)
+    }
+
+    /// Get the `RegulatedCoin.owner` field;
+    public fun owner<T>(c: &RegulatedCoin<T>): address {
+        c.owner
+    }
+
+    // === Necessary set of Methods (provide security guarantees and balance access) ===
+
+    /// Get an immutable reference to the Balance of a RegulatedCoin;
+    public fun borrow<T: drop>(_: T, coin: &RegulatedCoin<T>): &Balance<T> {
+        &coin.balance
+    }
+
+    /// Get a mutable reference to the Balance of a RegulatedCoin;
+    public fun borrow_mut<T: drop>(_: T, coin: &mut RegulatedCoin<T>): &mut Balance<T> {
+        &mut coin.balance
+    }
+
+    /// Author of the currency can restrict who is allowed to create new balances;
+    public fun zero<T: drop>(_: T, owner: address, ctx: &mut TxContext): RegulatedCoin<T> {
+        RegulatedCoin { id: TxContext::new_id(ctx), balance: Balance::zero(), owner }
+    }
+
+    /// Build a transferable `RegulatedCoin` from a `Balance`;
+    public fun from_balance<T: drop>(
+        _: T, balance: Balance<T>, owner: address, ctx: &mut TxContext
+    ): RegulatedCoin<T> {
+        RegulatedCoin { id: TxContext::new_id(ctx), balance, owner }
+    }
+
+    /// Destroy `RegulatedCoin` and return its `Balance`;
+    public fun into_balance<T: drop>(_: T, coin: RegulatedCoin<T>): Balance<T> {
+        let RegulatedCoin { balance, owner: _, id } = coin;
+        Sui::ID::delete(id);
+        balance
+    }
+
+    // === Optional Methods (can be used for simpler implementation of basic operations) ===
+
+    /// Join Balances of a `RegulatedCoin` c1 and `RegulatedCoin` c2.
+    public fun join<T: drop>(witness: T, c1: &mut RegulatedCoin<T>, c2: RegulatedCoin<T>) {
+        Balance::join(&mut c1.balance, into_balance(witness, c2))
+    }
+
+    /// Subtract `RegulatedCoin` with `value` from `RegulatedCoin`.
+    ///
+    /// This method does not provide any checks by default and can possibly lead to mocking
+    /// behavior of `RegulatedCoin::zero()` when a value is 0. So in case empty balances
+    /// should not be allowed, this method should be additionally protected against zero value.
+    public fun split<T: drop>(
+        witness: T, c1: &mut RegulatedCoin<T>, owner: address, value: u64, ctx: &mut TxContext
+    ): RegulatedCoin<T> {
+        let balance = Balance::split(&mut c1.balance, value);
+        from_balance(witness, balance, owner, ctx)
+    }
+}
+
+/// ABC is a RegulatedCoin which:
+///
+/// - is managed account creation (only admins can create a new balance)
+/// - has a denylist for addresses managed by the coin admins
+/// - has restricted transfers which can not be taken by anyone except the recipient
+module ABC::ABC {
+    use RC::RegulatedCoin::{Self as RC, RegulatedCoin as RC};
```

Comment:
Super minor nit, maybe just `RCoin`? I dream we might use `Rc` for the same thing as "ref counted" in Rust. It could be a good generalization of what's going on for global storage values inside the VM. 
(Imagine `borrow_global<T>(...): Rc<T>` or something. Could be interesting)

Reply:
Agreed! Used RC to write less code during development. 🐞